[Ingest UI] WaffleMap

simianhacker opened the pull request:
After pulling in upstream I realized that this PR was also damaged in yesterday's VS Code/Git/Merge snafu. The original PR was located at #21744

---

This PR adds the basic functionality for the Waffle Map to the Infra UI. Currently there is nothing configurable via the UI. The options are hard coded in the `withMap` container query, this will change as the UI features are developed to support it. 

![image](https://user-images.githubusercontent.com/41702/43792307-1949e7ae-9a2d-11e8-9e61-748a53b578e4.png)

How to test in browser:

1. Point Kibana at the `demo-stack` Kibana instance via `config/kibana.dev.yml` (reach out to @simianhacker for actual configuration) 
2. Click on infra icon. You should see something very close to above.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/9225/)

weltenwort commented on x-pack/plugins/infra/public/containers/map/with_map.ts:
```diff
@@ -3,22 +3,25 @@
 } from '../../../common/graphql/types';
+import { InfraWaffleMapGroup } from '../../lib/lib';
+import { nodesToWaffleMap } from './nodes_to_wafflemap';
 import { query } from './query';
```
The query imported here is named `Query`. That's a very conflict-prone name, especially since it will end up in the generated types. How about we call it something like

```typescript
export const mapQuery = gql`
  query MapQuery(...) {
    ...
  }
`;
```

Then the generated query namespace would be `MapQuery` in `common/graphql/types.ts` (with the query result type as `MapQuery.Query` and the variables as `MapQuery.Variables`, see below).

weltenwort commented on x-pack/plugins/infra/public/containers/map/with_map.ts:
```diff
@@ -52,8 +55,15 @@ export const withMap = graphql<
-  props: ({ data }) => ({ map: (data && data.source && data.source.map) || {} }),
+  props: ({ data }) => {
+    const nodes = get(data, 'source.map.nodes', []);
+    return { map: nodesToWaffleMap(nodes) };
```
According to the schema `source.map.nodes` can be `null`, which means the `nodesToWaffleMap` function fails at runtime when the query does not yield any nodes:

![image](https://user-images.githubusercontent.com/973741/43953184-796d51ee-9c98-11e8-9fa9-ef00866f04e6.png)

The compiler would have warned about this if instead of the hand-written `Response`, the generated `Query.Query` (or `MapQuery.Query` after the proposed renaming above) from `common/graphql/types.ts` had been used.

How about fixing that by

1. use the generated query response type and
2. change the schema and resolver of `InfraResponse` to always return an array for `nodes` (empty when no nodes are found)?

The same goes for the `Variables`. Let's use the generated `Query.Variables` (or `MapQuery.Variables`).

simianhacker commented:
@weltenwort I implemented the changes you requested. The `if` statements in the `withMap` HOC are a little meh but technically it's sound. I also had to add an ID to the nodes being returned because GraphQL seem to use that to do deduplication or something beyond my understanding the resulted all the nodes ending up in one path.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/9307/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/19/)

weltenwort commented on x-pack/plugins/infra/server/lib/adapters/nodes/lib/extract_group_paths.ts:
```diff
@@ -18,34 +18,26 @@ export function extractGroupPaths(
     (acc: InfraNode[], bucket: InfraBucket, index: number): InfraNode[] => {
       const nodeItem = createNodeItem(options, node, bucket);
       const key: string = String(bucket.key || index);
       if (secondGroup) {
```
Should this be `node.path_1` for consistency?

skh commented on x-pack/plugins/infra/public/components/waffle/group_of_groups.tsx:
```diff
@@ -0,0 +1,53 @@
+  );
+};
+
+export const Container = styled.div`
```
Could we use a less generic name for this?

simianhacker replied:
What name would have made more sense here? 

Keep in mind I think that name is only visible in this file. It doesn't leak outside the file anywhere else, those styled components are isolated to the file. If you look at the page with the React dev tool it shows up under a different name. It was a mistake to `export` those containers, force of habit :D

skh replied:
We could follow the naming convention you used in the other components, so it would be a `GroupOfGroupsContainer` in this case. (See also the review comment above.)

skh commented on x-pack/plugins/infra/public/components/waffle/index.tsx:
```diff
@@ -0,0 +1,94 @@
+  };
+}
+
+const Container = styled.div`
```
Could we use a less generic name for this?

simianhacker replied:
See above

skh commented on x-pack/plugins/infra/public/components/waffle/node.tsx:
```diff
@@ -0,0 +1,126 @@
+  };
+}
+
+export const Container = styled.div`
```
Could we use a less generic name for this?

simianhacker replied:
See above

skh commented on x-pack/plugins/infra/public/components/waffle/index.tsx:
```diff
@@ -0,0 +1,94 @@
+import { InfraWaffleData, InfraWaffleMapGroup, InfraWaffleOptions } from '../../lib/lib';
+import { applyWaffleMapLayout } from './lib/apply_wafflemap_layout';
+
+interface WafleMapProps {
```
Typo: `WaffleMapProps`

simianhacker replied:
Good cactch

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/36/)

simianhacker commented:
@skh I updated the names to be more specific. I also removed all the exports since I never intended those to be used outside the context of that component. 

@weltenwort We should be ready

weltenwort commented:
ah, it's running now :tada:

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/29/)

weltenwort commented:
looks like unrelated test failures related to the region map

simianhacker commented:
Merging... since we are just merging into our own feature branch, if it something we broke we can fix it later. Moving on.

weltenwort commented:
:tada:
